cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,194 @@
+   const int warp_id = tid / warp_size;
+   const size_type warps_per_grid = gridDim.x * block_size / warp_size;
+
+   // begin/end indices for the column data
+   size_type begin = 0;
+   size_type end = lhs.size();   
+   // warp indices.  since 1 warp == 32 threads == sizeof(bit_mask_t) * 8,
+   // each warp will process one (32 bit) of the validity mask via 
+   // __ballot_sync()
+   size_type warp_begin = cudf::word_index(begin);
+   size_type warp_end = cudf::word_index(end-1);
+
+   // lane id within the current warp
+   constexpr size_type leader_lane{0};
+   const int lane_id = threadIdx.x % warp_size;
+
+   size_type warp_valid_count{0};
+  
+   // current warp.
+   size_type warp_cur = warp_begin + warp_id;
+   size_type index = tid;
+   while(warp_cur <= warp_end){
+      bool in_range = (index >= begin && index < end);
+
+      bool valid = true;
+      if(has_validity){
+         valid = in_range && (filter(index) ? lhs.is_valid(index) : rhs.is_valid(index));
+      }
+
+      // do the copy if-else
+      if(in_range){ 
+         out.element<T>(index) = filter(index) ? lhs.element<T>(index) : rhs.element<T>(index);
+      }
+
+      // update validity
+      if(has_validity){
+         // the final validity mask for this warp
+         int warp_mask = __ballot_sync(0xFFFF'FFFF, valid && in_range);
+         // only one guy in the warp needs to update the mask and count
+         if(lane_id == 0){
+            out.set_mask_word(warp_cur, warp_mask);
+            warp_valid_count += __popc(warp_mask);
+         }
+      }
+
+      // next grid
+      warp_cur += warps_per_grid;
+      index += block_size * gridDim.x;       
+   }
+
+   if(has_validity){
+      // sum all null counts across all warps
+      size_type block_valid_count = single_lane_block_sum_reduce<block_size, leader_lane>(warp_valid_count);
+      // block_valid_count will only be valid on thread 0
+      if(threadIdx.x == 0){
+         // using an atomic here because there are multiple blocks doing this work
+         atomicAdd(valid_count, block_valid_count);
+      }      
+   }
+}
+
+
+
+/* --------------------------------------------------------------------------*/
+/**
+* @brief Functor called by the `type_dispatcher` in order to perform a copy if/else
+*        using a filter function to select from lhs/rhs columns.
+*/
+/* ----------------------------------------------------------------------------*/
+struct copy_if_else_functor {
+   template <typename T, typename Filter>
+   std::unique_ptr<column> operator()(column_view const& lhs,
+                                    column_view const& rhs,
+                                    Filter filter,
+                                    rmm::mr::device_memory_resource *mr,
+                                    cudaStream_t stream)
+   {
+      // output
+      std::unique_ptr<column> out = experimental::allocate_like(lhs, lhs.size(), experimental::mask_allocation_policy::RETAIN, mr);
```

Comment:
This is incorrect. If `lhs.nullable() == false` and `rhs.has_nulls() == true`, then the output should be nullable.

Reply:
The mask allocation policy should be something like:
```
auto policy = (lhs.has_nulls() or rhs.has_nulls()) ? ALWAYS : NEVER;
```

Reply:
Should it be has_nulls() or .nullable()?   In the case where you've got validity data but everything is valid, has_nulls() would cause us to drop the validity mask from the output, which may be slightly unexpected from the callers perspective.

Reply:
My opinion is that a column without a null mask is semantically equivalent to a column with a null mask with all bits set. I generally opt to be more restrictive and check `null_count()` in situations like these.

Reply:
Sounds good.

Reply:
Actually, spoke too soon.  Here's a case that this makes a little weird.

```
// a, b, c all have validity masks and no nulls.
// out will end up with no validity mask.
out = copy_if_else(a, b, c);

// contains the proper results, but -also- has validity mask + no nulls.
column expected;

// this fails because out doesn't have a validity mask.  
expect_columns_equal(out, expected);
```

So at the very least, expect_columns_equal() is not treating these two as semantically equivalent.

Reply:
Yeah, this came up before and I asked someone to update `expect_columns_equal` to match the behavior you described. I don't think it was ever done. Feel free to do as part of this PR.